feathr_project/feathr/_synapse_submission.py

```diff
@@ -347,14 +348,40 @@ def download_file(self, target_adls_directory: str, local_dir_cache: str):
         parse_result = urlparse(target_adls_directory)
         directory_client = self.file_system_client.get_directory_client(
             parse_result.path)
+
         # returns the paths to all the files in the target director in ADLS
-        adls_paths = [file_path.name.split("/")[-1] for file_path in self.file_system_client.get_paths(
-            path=parse_result.path) if not file_path.is_directory][1:]
-        # need to generate list of local paths to write the files to
+        # get all the paths that are not under a directory
+        result_paths = [basename(file_path.name) for file_path in self.file_system_client.get_paths(
+            path=parse_result.path, recursive=False) if not file_path.is_directory]
+        
+        # get all the paths that are directories and download them
+        result_folders = [file_path.name for file_path in self.file_system_client.get_paths(
+            path=parse_result.path) if file_path.is_directory]
+
+        # list all the files under the certain folder, and download them preserving the hierarchy
+        for folder in result_folders:
+            folder_name = basename(folder)
+            file_in_folder = [os.path.join(folder_name, basename(file_path.name)) for file_path in self.file_system_client.get_paths(
+            path=folder, recursive=False) if not file_path.is_directory]
+            local_paths = [os.path.join(local_dir_cache, file_name)
+                       for file_name in file_in_folder]
+            self._download_file_list(local_paths, file_in_folder, directory_client)
+
+        # download files that are in the result folder
         local_paths = [os.path.join(local_dir_cache, file_name)
-                       for file_name in adls_paths]
-        for idx, file_to_write in enumerate(tqdm(adls_paths,desc="Downloading result files: ")):
+                       for file_name in result_paths]
+        self._download_file_list(local_paths, result_paths, directory_client)
+
+        logger.info('Finish downloading files from {} to {}.',
+                    target_adls_directory,local_dir_cache)
```

Comment:
Could you explain why is this different from Databricks?

Reply:
Mostly because of different underlying implementations. For databricks we are directly calling the REST APIs but for Synapse there's a Python SDK. They use different ways for file/path download hence we need to adjust accordingly.